### PR TITLE
fix: reduce token list rerenders

### PR DESCRIPTION
**Pull request by guanbinrui**
## Description

Closes # (NO_ISSUE)

### Before

https://user-images.githubusercontent.com/52657989/162123010-bd7e9ef5-ba1b-4406-a260-5773659ab319.mp4 

### After

https://user-images.githubusercontent.com/52657989/162123014-1f0bef61-5001-431e-ade2-4e2d5532c265.mp4

## Type of change

- [ ] Documentation
- [ ] Code refactoring (Restructuring existing code w/o changing its observable behavior)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (a fix or feature that would make something no longer possible to do/require old user must upgrade their Mask Network to this new version)

### CODE REVIEW

File: packages/shared/src/UI/components/ERC20TokenList/index.tsx [hunk 1]
```diff
@@ -87,8 +88,9 @@ export const ERC20TokenList = memo<ERC20TokenListProps>((props) => {
             (!excludeTokens.length || !excludeTokens.some(currySameAddress(token.address))),
     )
 
-    const renderTokens = uniqBy([...tokens, ...filteredTokens, ...(searchedToken ? [searchedToken] : [])], (x) =>
-        x.address.toLowerCase(),
+    const renderTokens = uniqBy(
+        [...tokens, ...filteredTokens, ...(searchedToken ? [searchedToken] : EMPTY_LIST)],
```
Comment (UncleBill) [inline packages/shared/src/UI/components/ERC20TokenList/index.tsx:92]: No need here.

File: packages/shared/src/UI/components/ERC20TokenList/index.tsx [hunk 2]
```diff
@@ -101,17 +103,24 @@ export const ERC20TokenList = memo<ERC20TokenListProps>((props) => {
         chainId,
     )
 
-    const renderAssets =
-        !account || !!assetsError || assetsLoading || searchedTokenLoading
-            ? [...renderTokens]
-                  .sort(makeSortTokenFn(chainId, { isMaskBoost: true }))
-                  .map((token) => ({ token: token, balance: null }))
+    const renderAssets = useMemo(() => {
```
Comment (UncleBill) [inline packages/shared/src/UI/components/ERC20TokenList/index.tsx:106]: Memo all, or memo none. Should consider memo other data as well.

Comment (UncleBill) [reply packages/shared/src/UI/components/ERC20TokenList/index.tsx:106]: should memo object passed to`SearchFieldProps` as well.

```tsx
        <SearchableList<Asset>
            SearchFieldProps={{
                placeholder: t.erc20_token_list_placeholder(),
                ...props.SearchTextFieldProps,
            }}

```

File: packages/theme/src/Components/SearchableList/SearchableList.tsx [hunk 3]
```diff
@@ -72,7 +73,7 @@ export function SearchableList<T>({
                 shouldSort: true,
                 threshold: 0.45,
                 minMatchCharLength: 1,
-                keys: searchKey ?? Object.keys(data.length > 0 ? data[0] : []),
+                keys: searchKey ?? Object.keys(data.length > 0 ? data[0] : EMPTY_LIST),
```
Comment (UncleBill) [inline packages/theme/src/Components/SearchableList/SearchableList.tsx]: No need to use EMP_LIST here
```suggestion
                keys: searchKey ?? data.length > 0 ? Object.keys(data[0]) : [],
```

File: packages/web3-shared/evm/hooks/useERC20TokensDetailedFromTokenLists.ts [hunk 4]
```diff
@@ -45,10 +46,10 @@ export function useERC20TokensDetailedFromTokenLists(
         if (!keyword) return allToken
 
         return [
-            ...(EthereumAddress.isValid(keyword) ? allToken.filter(currySameAddress(keyword)) : []),
+            ...(EthereumAddress.isValid(keyword) ? allToken.filter(currySameAddress(keyword)) : EMPTY_LIST),
```
Comment (UncleBill) [inline packages/web3-shared/evm/hooks/useERC20TokensDetailedFromTokenLists.ts]: No need to use EMP_LIST here

File: packages/shared/src/UI/components/ERC20TokenList/index.tsx [hunk 5]
```diff
@@ -101,17 +103,24 @@ export const ERC20TokenList = memo<ERC20TokenListProps>((props) => {
...
             : keyword
             ? assets
             : [...assets].sort(makeSortAssertFn(chainId, { isMaskBoost: true }))
+    }, [
+        chainId,
+        keyword,
+        assetsLoading,
+        assets
+            .map((x) => x.token.address)
+            .sort()
```
Comment (UncleBill) [inline packages/shared/src/UI/components/ERC20TokenList/index.tsx:118]: Should not sort here, right?

File: packages/shared/src/UI/components/ERC20TokenList/index.tsx (hunk 1 above)
Comment (UncleBill) [reply packages/shared/src/UI/components/ERC20TokenList/index.tsx:92]: But memo would be better.

File: packages/shared/src/UI/components/ERC20TokenList/index.tsx (hunk 2 above)
Comment (guanbinrui) [reply packages/shared/src/UI/components/ERC20TokenList/index.tsx:106]: No necessary. They don't change actively.

File: packages/shared/src/UI/components/ERC20TokenList/index.tsx (hunk 5 above)
Comment (guanbinrui) [reply packages/shared/src/UI/components/ERC20TokenList/index.tsx:118]: I will keep this.

File: packages/shared/src/UI/components/ERC20TokenList/index.tsx (hunk 1 above)
Comment (guanbinrui) [reply packages/shared/src/UI/components/ERC20TokenList/index.tsx:92]: I will keep this.

**Review by UncleBill — APPROVED** [review]

**Review by Jack-Works — APPROVED** [review]

**Pull request merged**